### ISSUE TITLE
Consider implementing `goodpractice` advice

### ISSUE THREAD

**Issue by gerasy1987** [integrated-inferences/CausalQueries#295]
`goodpractice` package provides function `gp()` that produces advice on good practice in package structure/coding. Unlike `autotest` and `pkgcheck` packages it actually works at this point without producing some obscure errors. Below is the output of calling `goodpractice::gp(.)` which all sounds reasonable to me and I would consider addressing at least some of it.

```
── GP CausalQueries ────────────────────────────────────────────────────────────────────────────

It is good practice to

  ✖ write unit tests for all functions, and all package code in general. 9%
    of code lines are covered by test cases.

    R/clean_params.R:17:NA
    R/clean_params.R:18:NA
    R/clean_params.R:27:NA
    R/clean_params.R:54:NA
    R/clean_params.R:56:NA
    ... and 2701 more lines

  ✖ write short and simple functions. These functions have high cyclomatic
    complexity (>50): make_par_values (61). You can make them easier to reason about
    by encapsulating distinct steps of your function into subfunctions.
  ✖ not use "Depends" in DESCRIPTION, as it can cause name clashes, and poor
    interaction with other packages. Use "Imports" instead.
  ✖ add a "URL" field to DESCRIPTION. It helps users find information about
    your package online. If your package does not have a homepage, add an URL to
    GitHub, or the CRAN package package page.
  ✖ add a "BugReports" field to DESCRIPTION, and point it to a bug tracker.
    Many online code hosting services provide bug trackers for free,
    https://github.com, https://gitlab.com, etc.
  ✖ use '<-' for assignment instead of '='. '<-' is the standard, and R users
    and developers are used it and it is easier to read your code for them if you use
    '<-'.

    R/make_data.R:145:4
    R/make_data.R:158:5
    R/map_query_to_causal_type.R:58:18
    R/query_helpers.R:20:15
    R/query_helpers.R:46:15
    ... and 8 more lines

  ✖ avoid long code lines, it is bad for readability. Also, many people
    prefer editor windows that are about 80 characters wide. Try make your lines
    shorter than 80 characters

    R/CausalQueries-package.R:5:81
    R/clean_params.R:1:81
    R/clean_params.R:3:81
    R/clean_params.R:4:81
    R/clean_params.R:5:81
    ... and 615 more lines

  ✖ avoid sapply(), it is not type safe. It might return a vector, or a list,
    depending on the input data. Consider using vapply() instead.

    R/data_helpers.R:238:18
    R/get_ambiguities_matrix.R:41:20
    R/get_ambiguities_matrix.R:54:10
    R/get_parents.R:13:3
    R/helpers.R:15:13
    ... and 54 more lines

  ✖ avoid 1:length(...), 1:nrow(...), 1:ncol(...), 1:NROW(...) and
    1:NCOL(...) expressions. They are error prone and result 1:0 if the expression on
    the right hand side is zero. Use seq_len() or seq_along() instead.

    R/data_helpers.R:238:25
    R/get_ambiguities_matrix.R:41:27
    R/get_ambiguities_matrix.R:54:17
    R/get_nodal_types.R:67:12
    R/get_nodal_types.R:69:12
    ... and 34 more lines

  ✖ not import packages as a whole, as this can cause name clashes between
    the imported packages. Instead, import only the specific functions you need.
  ✖ fix this R CMD check NOTE: GNU make is a SystemRequirements.
  ✖ fix this R CMD check NOTE: installed size is 50.1Mb sub-directories of
    1Mb or more: libs 49.5Mb
```

**Comment by till-tietz** [comment]
Thanks for flagging @gerasy1987  will run `gp()` locally and fix issues. Some key ones that are easy + worthwhile are: 
* type safety 
* range based loops 
* readability 
* assignment operators 

Please weight in here @macartan  @gerasy1987  @lilymedina if you think further fixes would be useful for the next CRAN update prior to JSS submit. 

Improving test coverage is not in scope atm I think. @macartan how do you feel about the current state of the tests? 


Will link a branch to this issue and update on implemented fixes in the comments.

**Comment by gerasy1987** [comment]
@till-tietz I agree on the priorities; all their direct coding suggestions seem straightforward to address and generally good. Also, the `gp()` function creates tags in the package recognized by RStudio, making it easy to find their suggestions' points.

I'm also wondering if their suggestions for `DEXCRIPTION` file and the first note from `R CMD check` can be quickly addressed.

Also, as a side note, `goodpractice` is called by `pkgcheck` functions, so if the JSS reviewers look at the package, they might try to run this anyway.

**Comment by macartan** [comment]
I agree; these all seem straight forward

I think our set of tests are pretty good; however it's not great if our
coverage is low; before I remember we got it up to 100% and it was quite a
useful exercise  and helped us find little bugs

now 2701 sounds like an awful lot! but it could easily be that lots of
these get run together once a particular option is put in

**Comment by macartan** [comment]
I'm having difficulties getting results from covr but if we have a list of functions that have especially poor coverage I am happy to write some tests

**Comment by till-tietz** [comment]
@macartan I'll get some test coverage results and will open a testing issue noting key functions

**Comment by macartan** [comment]
Sounds great

I imagine the many options in set priors and set parameters might be an
issue here and it might be possible to use expand grid or something Tk make
list of arguments and run bunch of combinations

If any functions aren’t covered at all then maybe they are obsolete!

**Comment by gerasy1987** [comment]
@macartan @till-tietz `gp()` function I showed the output above runs `covr()` under the hood and creates tags for where the coverage is missing, which makes it easier to find what tests to write. I can help with this, too! Also, as we do that, it might be a good idea to rewrite examples using native pipelines, but I can create a separate issue for that.

**Comment by macartan** [comment]
I know covr should work for me but for some reason I have some permissions
issues with installation where I have to restart R each time and clear
objects and these seems tp prevent covr from working; I got the bp report
without the covr components

by native you mean |> rather than %>%? we can do find replace for that
maybe though there might be issues with missing ()

**Comment by gerasy1987** [comment]
@macartan on the native pipeline replacement, yes I meant `%>%` to `|>` transition at least in the examples. Find-replace is very error prone given the `()` issue and also that `|>` uses different conventions for where it plugs the LHS expression, so my proposal was that whoever writes the tests for a specific function can also go through examples and do this replacement carefully for that function.

**Comment by till-tietz** [comment]
@macartan & @gerasy1987 ... agreed on transitioning from %>% to |> . |> does not support dot nation for passing args and generally requires spelling out arg names when not passing to the first function argument. Replacement on a case by case basis is likely safer here.

**Comment by till-tietz** [comment]
@macartan @gerasy1987 @lilymedina goodpractice fixes are implemented wherever possible (some warnings/ suggestions apply to R code that is generated by STAN so I did not apply fixes there). 

We are passing rcmd checks + runs on examples and tests look good. Some changes to enforce type safety of sapply calls were extensive and tricky (particularly in instances where we previously relied on non-type safe behavior of sapply). 
It would be good to pay particular attention to potential edge cases when doing your reviews. Key focus areas are:

* `make_par_values()` (particularly with confounding)
* `st_within()` calls that may return 0 length vectors as I have changed the handling of this case (particularly relevant for malformed queries/statements in `map_query_to_causal_type()`, `map_query_to_nodal_type()`, `expand_wildcard()` )
